share/getters/cascade.go

```diff
@@ -0,0 +1,207 @@
+package getters
+
+import (
+	"context"
+	"time"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/multierr"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/nmt/namespace"
+	"github.com/celestiaorg/rsmt2d"
+)
+
+var cascadeTracer = otel.Tracer("getters/cascade")
+
+var _ share.Getter = (*CascadeGetter)(nil)
+
+// CascadeGetter implements custom share.Getter that composes multiple Getter implementations in
+// "cascading" order.
+//
+// See cascade func for details on cascading.
+type CascadeGetter struct {
+	interval time.Duration
+
+	getters []share.Getter
+}
+
+// NewCascadeGetter instantiates a new CascadeGetter from given share.Getters with given interval.
+func NewCascadeGetter(getters []share.Getter, interval time.Duration) *CascadeGetter {
+	return &CascadeGetter{
+		interval: interval,
+		getters:  getters,
+	}
+}
+
+// GetShare gets a share from any of registered share.Getters in cascading order.
+func (cg *CascadeGetter) GetShare(ctx context.Context, root *share.Root, row, col int) (share.Share, error) {
+	ctx, span := cascadeTracer.Start(ctx, "get-share", trace.WithAttributes(
+		attribute.String("root", root.String()),
+		attribute.Int("row", row),
+		attribute.Int("col", col),
```

Comment:
Hmm:
* Each Getter would have its span anyway, with a custom name describing the implementation, like [here](https://github.com/celestiaorg/celestia-node/pull/1628/files#diff-830b9c493d28c2431bda3668469f2d55db8e9112f3211c1f12f264570baf4780R41-R44). In this PR we have a personal tracer for cascade getter which would mark it as so, so wouldn't be hard to debug.
* I am not a fan of extending interfaces with statically returned values. We could make it through anonymous interface(check if a Getter implement Name() - use it), but I still don't see why according to the point above

Reply:
Getter spans don't have info of getter name. In the provided pr, there is arbitary getter_idx attribute, that could be confusing to identify what getter implementation was actually in use. Wrapping span or log with getter name could see great benefit for debugging.

It could be also to use extended interface like.
```
type NamedGetter interface{
    Name() string
    Getter
}

Or even log/trace wrap getter... 
But thats unnecessary overcomplicating things, when adding extra static method is super easy.